File: sodp-go/acl.go
package sodp

import (
	"encoding/json"
	"fmt"
	"os"
	"strings"
)

// aclRule is one entry in the ACL JSON file.
// Pattern uses dot-separated segments: literals, {sub} (captures JWT sub),
// and a trailing * that matches one or more remaining segments.
//
// Example rule file:
//
//	[
//	  {"pattern":"private.{sub}","read":["{sub}"],"write":["{sub}"]},
//	  {"pattern":"admin.*","read":["role:admin"],"write":["role:admin"]},
//	  {"pattern":"public.*","read":["*"],"write":["role:writer"]}
//	]
type aclRule struct {
	Pattern string   `json:"pattern"`
	Read    []string `json:"read"`
	Write   []string `json:"write"`
}

// ACL is a parsed set of access-control rules.
type ACL struct {
	rules []aclRule
}

// loadACL reads and parses an ACL JSON file.
func loadACL(path string) (*ACL, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, fmt.Errorf("sodp: acl: read %s: %w", path, err)
	}
	var rules []aclRule
	if err := json.Unmarshal(data, &rules); err != nil {
		return nil, fmt.Errorf("sodp: acl: parse: %w", err)
	}
	return &ACL{rules: rules}, nil
}

// Authorize returns true if the session may perform action ("read" or "write")
// on key. Rules are evaluated in order; the first matching rule is decisive.
// If no rule matches, access is allowed (permissive default).
func (a *ACL) Authorize(sess *Session, key, action string) bool {
	for _, rule := range a.rules {
		captures, ok := matchACLPattern(rule.Pattern, key)
		if !ok {
			continue
		}
		var perms []string
		if action == "write" {
			perms = rule.Write
		} else {
			perms = rule.Read
		}
		return checkPerms(perms, sess, captures)
	}
	return true // no rule matched → allow
}

// matchACLPattern matches a dot-separated pattern against key.
// Returns (captures, true) on match, (nil, false) on miss.
func matchACLPattern(pattern, key string) (map[string]string, bool) {
	pSegs := strings.Split(pattern, ".")
	kSegs := strings.Split(key, ".")
	captures := make(map[string]string, 2)

	for i, p := range pSegs {
		if p == "*" {
			return captures, i < len(kSegs) // wildcard: must consume ≥1 remaining segment
		}
		if i >= len(kSegs) {
			return nil, false
		}
		if len(p) >= 3 && p[0] == '{' && p[len(p)-1] == '}' {
			captures[p[1:len(p)-1]] = kSegs[i]
		} else if p != kSegs[i] {
			return nil, false
		}
	}
	return captures, len(pSegs) == len(kSegs)
}

// checkPerms returns true if any permission in the list grants the session access.
//
// Permission formats:
//   - "*"         — any authenticated (or anonymous) session
//   - "{sub}"     — session's JWT sub must equal the captured {sub} segment
//   - "role:X"    — session must have role X in claims["roles"] or claims["role"]
//   - "group:X"   — session must have group X in claims["groups"]
//   - "perm:X"    — session must have permission X in claims["permissions"]
//   - "claim:K:V" — claims[K] must equal or contain V
//   - literal     — session's sub must equal this string exactly
func checkPerms(perms []string, sess *Session, captures map[string]string) bool {
	for _, perm := range perms {
		switch {
		case perm == "*":
			return true
		case perm == "{sub}":
			if sub, ok := captures["sub"]; ok && sub == sess.Sub {
				return true
			}
		case strings.HasPrefix(perm, "role:"):
			role := perm[5:]
			if aclHasClaim(sess.Claims, "roles", role) || aclHasClaim(sess.Claims, "role", role) ||
				aclHasClaim(sess.Claims, "realm_access.roles", role) {
				return true
			}
		case strings.HasPrefix(perm, "group:"):
			if aclHasClaim(sess.Claims, "groups", perm[6:]) {
				return true
			}
		case strings.HasPrefix(perm, "perm:"):
			if aclHasClaim(sess.Claims, "permissions", perm[5:]) {
				return true
			}
		case strings.HasPrefix(perm, "claim:"):
			// "claim:KEY:VALUE"
			rest := perm[6:]
			if colon := strings.IndexByte(rest, ':'); colon > 0 {
				if aclHasClaim(sess.Claims, rest[:colon], rest[colon+1:]) {
					return true
				}
			}
		default:
			if perm == sess.Sub {
				return true
			}
		}
	}
	return false
}

// aclHasClaim checks whether claims[key] equals value or is a list containing value.
// Supports dot-traversal for nested claims (e.g. "realm_access.roles").
func aclHasClaim(claims map[string]any, key, value string) bool {
	var v any
	if dot := strings.IndexByte(key, '.'); dot >= 0 {
		parent, ok := claims[key[:dot]].(map[string]any)
		if !ok {
			return false
		}
		v = parent[key[dot+1:]]
	} else {
		v = claims[key]
	}
	switch cv := v.(type) {
	case string:
		return cv == value
	case []any:
		for _, item := range cv {
			if s, ok := item.(string); ok && s == value {
				return true
			}
		}
	}
	return false
}

File: server-ts/src/acl.ts
// Pattern-based ACL matching the Rust server grammar.
//
// Rule format (acl.json):
//   [{ "pattern": "user.{sub}.*", "read": ["{sub}"], "write": ["{sub}"] }]
//
// Pattern segments (dot-separated):
//   literal  — exact match ("game", "scores")
//   {sub}    — captures the JWT sub claim; one segment only
//   *        — matches one or more remaining segments (only valid at end)
//
// Permission expressions:
//   "*"        — anyone (including unauthenticated)
//   "{sub}"    — claims.sub must equal the captured {sub}
//   "role:X"   — claims.roles includes X
//   "group:X"  — claims.groups includes X
//   "perm:X"   — claims.perms includes X

import { readFileSync } from 'fs'
import type { SodpClaims } from './jwt'

export interface AclRule {
  pattern: string
  read: string[]
  write: string[]
}

interface CompiledRule {
  segments: Array<{ type: 'literal'; value: string } | { type: 'capture' } | { type: 'wildcard' }>
  read: string[]
  write: string[]
}

function compileRule(rule: AclRule): CompiledRule {
  const parts = rule.pattern.split('.')
  const segments: CompiledRule['segments'] = parts.map((p) => {
    if (p === '*') return { type: 'wildcard' }
    if (p === '{sub}') return { type: 'capture' }
    return { type: 'literal', value: p }
  })
  return { segments, read: rule.read, write: rule.write }
}

function matchPattern(
  compiled: CompiledRule,
  keySegments: string[],
): { matched: boolean; captures: { sub?: string } } {
  const { segments } = compiled
  const captures: { sub?: string } = {}

  let ki = 0
  for (let si = 0; si < segments.length; si++) {
    const seg = segments[si]
    if (seg.type === 'wildcard') {
      // Wildcard must be last and must consume at least one key segment.
      if (si !== segments.length - 1) return { matched: false, captures }
      if (ki >= keySegments.length) return { matched: false, captures }
      return { matched: true, captures }
    }
    if (ki >= keySegments.length) return { matched: false, captures }
    if (seg.type === 'literal') {
      if (keySegments[ki] !== seg.value) return { matched: false, captures }
    } else {
      // capture
      captures.sub = keySegments[ki]
    }
    ki++
  }

  return { matched: ki === keySegments.length, captures }
}

function checkPermission(
  perms: string[],
  claims: SodpClaims | null,
  captures: { sub?: string },
): boolean {
  for (const p of perms) {
    if (p === '*') return true
    if (!claims) continue
    if (p === '{sub}' && captures.sub !== undefined && claims.sub === captures.sub) return true
    if (p.startsWith('role:')  && claims.roles.includes(p.slice(5))) return true
    if (p.startsWith('group:') && claims.groups.includes(p.slice(6))) return true
    if (p.startsWith('perm:')  && claims.perms.includes(p.slice(5))) return true
    // Literal sub match
    if (claims && p === claims.sub) return true
  }
  return false
}

export class AclRegistry {
  private readonly rules: CompiledRule[]

  constructor(rules: AclRule[]) {
    this.rules = rules.map(compileRule)
  }

  static fromFile(path: string): AclRegistry {
    const raw = readFileSync(path, 'utf-8')
    return new AclRegistry(JSON.parse(raw) as AclRule[])
  }

  // Returns true if the given claims may read the key.
  // If no rule matches, access is allowed (permissive default — same as Rust server).
  canRead(key: string, claims: SodpClaims | null): boolean {
    return this.check(key, claims, 'read')
  }

  canWrite(key: string, claims: SodpClaims | null): boolean {
    return this.check(key, claims, 'write')
  }

  private check(key: string, claims: SodpClaims | null, mode: 'read' | 'write'): boolean {
    const keySegments = key.split('.')
    for (const rule of this.rules) {
      const { matched, captures } = matchPattern(rule, keySegments)
      if (!matched) continue
      const perms = mode === 'read' ? rule.read : rule.write
      return checkPermission(perms, claims, captures)
    }
    // No rule matched → permissive
    return true
  }
}
